api/v3/Order.php

```diff
@@ -73,62 +73,73 @@ function _civicrm_api3_order_get_spec(array &$params) {
         $order->setLineItem($lineItem, $lineIndex);
-      }
-
-      $entityParams = $lineItems['params'] ?? NULL;
-
-      if ($entityParams && $order->getLineItemEntity($lineIndex) !== 'contribution') {
-        switch ($order->getLineItemEntity($lineIndex)) {
-          case 'participant':
-            if (isset($entityParams['participant_status_id'])
-              && (!CRM_Event_BAO_ParticipantStatusType::getIsValidStatusForClass($entityParams['participant_status_id'], 'Pending'))) {
-              throw new CiviCRM_API3_Exception('Creating a participant via the Order API with a non "pending" status is not supported');
-            }
-            $entityParams['participant_status_id'] = $entityParams['participant_status_id'] ?? 'Pending from incomplete transaction';
-            $entityParams['status_id'] = $entityParams['participant_status_id'];
-            $entityParams['skipLineItem'] = TRUE;
-            $entityResult = civicrm_api3('Participant', 'create', $entityParams);
-            // @todo - once membership is cleaned up & financial validation tests are extended
-            // we can look at removing this - some weird handling in removeFinancialAccounts
-            $params['contribution_mode'] = 'participant';
-            $params['participant_id'] = $entityResult['id'];
-            break;
-
-          case 'membership':
-            $entityParams['status_id'] = 'Pending';
-            if (!empty($params['contribution_recur_id'])) {
-              $entityParams['contribution_recur_id'] = $params['contribution_recur_id'];
-            }
-            $entityParams['skipLineItem'] = TRUE;
-            $entityResult = civicrm_api3('Membership', 'create', $entityParams);
-            break;
-
-        }
-
-        foreach ($lineItems['line_item'] as $innerIndex => $lineItem) {
-          $lineIndex = $index . '+' . $innerIndex;
-          $order->setLineItemValue('entity_id', $entityResult['id'], $lineIndex);
-        }
+        $order->addLineItemToEntityParameters($lineIndex, $index);
       }
     }
-    $priceSetID = $order->getPriceSetID();
-    $params['line_item'][$priceSetID] = $order->getLineItems();
   }
   else {
     $order->setPriceSetToDefault('contribution');
+    $order->setLineItem([
+      // Historically total_amount in this case could be tax
+      // inclusive if tax is also supplied.
+      // This is inconsistent with the contribution api....
+      'line_total' => (float) ($params['total_amount'] - ($params['tax_amount'] ?? 0)),
+      'financial_type_id' => (int) $params['financial_type_id'],
+    ], 0);
+  }
+  if (isset($params['total_amount'])) {
+    $currency = $params['currency'] ?? CRM_Core_Config::singleton()->defaultCurrency;
+    if (!CRM_Utils_Money::equals($params['total_amount'], $order->getTotalAmount(), $currency)) {
+      throw new CRM_Contribute_Exception_CheckLineItemsException();
+    }
   }
 
+  foreach ($order->getEntitiesToCreate() as $entityParams) {
```

Comment:
All of this will be moved to BAO_Order - the two things to sort out first are

1) the below loop alters $params - which should not be done in BAO_Order - to set contribution_mode - once https://github.com/civicrm/civicrm-core/pull/20650 is merged I can start try to remove that and
2) We will want BAO_Order to use v4 api internally - but we will need to do some wrangling to support v3 style entity params from v3 api